django/db/models/fields/tuple_lookups.py

```diff
@@ -196,14 +210,25 @@ def as_oracle(self, compiler, connection):
 
 class TupleIn(TupleLookupMixin, In):
     def get_prep_lookup(self):
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_is_collection_of_tuples_or_lists()
         self.check_rhs_elements_length_equals_lhs_length()
         return super(TupleLookupMixin, self).get_prep_lookup()
```

Comment:
This is really complicated. We're skipping `TupleLookupMixin.get_prep_lookup()` via `super` magic so we can call a different collection of helper methods.

I wonder if a better approach would be to add to `Tuplein`:

```python
    def check_rhs_is_tuple_or_list():
        super().check_rhs_is_tuple_or_list()
        self.check_rhs_is_collection_of_tuples_or_lists()

    def check_rhs_length_equals_lhs_length():
         # We don't call `super()` here because we want to completely
         # override the TupleLookupMixin version.
         self.check_rhs_elements_length_equals_lhs_length()
```

Then we don't need to override `get_prep_lookup` here.

Reply:
We could also inline `check_rhs_elements_length_equals_lhs_length` and/or `check_rhs_is_collection_of_tuples_or_lists` here.

Reply:
That works too, though I think it deserves a comment to explain why we can't just use `super`.

Reply:
I added a comment. I think this is fine for now, I'm resolving this conversation.